src/components/DataProvider/index.js

```diff
@@ -35,21 +36,81 @@ export default class DataProvider extends Component {
     yDomains: {},
   };
 
+  static getDerivedStateFromProps(nextProps, prevState) {
+    // Check if one of the series got removed from props
+    // If so, delete the respective keys in contextSeries and loaderconfig
+    // This is important so we don't cache the vales if it gets readded later
+    const { loaderConfig, contextSeries, yDomains } = prevState;
+    const { series } = nextProps;
+    const seriesKeys = {};
+    series.forEach(s => {
+      seriesKeys[s.id] = true;
+    });
+    const newContextSeries = { ...contextSeries };
+    const newLoaderConfig = { ...loaderConfig };
+    const newYDomains = { ...yDomains };
+    let shouldUpdate = false;
+    Object.keys(loaderConfig).forEach(key => {
+      if (!seriesKeys[key]) {
+        // Clean up
+        delete newContextSeries[key];
+        delete newLoaderConfig[key];
+        delete newYDomains[key];
+        shouldUpdate = true;
+      }
+    });
+    if (shouldUpdate) {
+      return {
+        loaderConfig: newLoaderConfig,
+        contextSeries: newContextSeries,
+        yDomains: newYDomains,
+      };
+    }
+    return null;
+  }
+
   async componentDidMount() {
-    await this.fetchData('MOUNTED');
+    await Promise.map(this.props.series, s => this.fetchData(s.id, 'MOUNTED'));
     if (this.props.updateInterval) {
       this.fetchInterval = setInterval(() => {
-        this.fetchData('INTERVAL');
+        Promise.map(this.props.series, s => this.fetchData(s.id, 'INTERVAL'));
       }, this.props.updateInterval);
     }
   }
 
+  async componentDidUpdate(prevProps) {
+    // If new series are present in prop,
+    // run the fetchData lifecycle for those series
+    const { series: prevSeries } = prevProps;
+    if (!prevSeries) {
+      return;
+    }
+    const { series, baseDomain } = this.props;
+    const { subDomain } = this.state;
+
+    const currentSeriesKeys = {};
+    series.forEach(s => {
+      currentSeriesKeys[s.id] = true;
+    });
+    const prevSeriesKeys = {};
+    prevSeries.forEach(p => {
+      prevSeriesKeys[p.id] = true;
+    });
+    const newSeries = series.filter(s => prevSeriesKeys[s.id] !== true);
+    await Promise.map(newSeries, async ({ id }) => {
+      await this.fetchData(id, 'MOUNTED');
+      if (!isEqual(subDomain, baseDomain)) {
+        await this.fetchData(id, 'UPDATE_SUBDOMAIN');
+      }
+    });
+  }
+
   componentWillUnmount() {
     clearInterval(this.fetchInterval);
     this.unmounted = true;
   }
 
-  getSeriesObjects() {
+  getSeriesObjects = () => {
```

Comment:
This is sort of a meta question, but why is `foo = () => {` preferred over `foo() {` ?

Reply:
Using arrow function notation is just the way of saving the lexical scope in `this` in order to call it from another class which became available as a part of [this proposal](https://github.com/tc39/proposal-class-fields). Consider you pass function as a prop to the child class inside of which you want to access parent's scope (i.e. to use props, setState, etc). If just to pass it as is, you will lose the scope, since `this` will be related to a child class. Consider the following:

```
const Child = ({handleClick}) => 
  <button onClick={handleClick}>Button</button>

class Parent extends Component {
  handleClick() {
    console.log(this.props);
  }

  render() {
    return 
  }
}
```


There are lot's of ways to workaround this behaviour, one of which is method arrow notation, another instead of passing method, wrap it in an arrow function and call (`() => this.method()`), another one is use `.bind()`, etc.

BTW there is actually no need to wrap `getSeriesObjects` here since it is called only inside the same class.

Reply:
I agree -- however `() => method()` will create a new function every time. There is no need here -- but I like the consistency.